### PR TITLE
Add boolean to varchar coercion for hive

### PR DESCRIPTION
**Pull request by findinpath**
## Description

Add boolean to varchar coercion for hive.

Enable the Hive users to retrieve the information from their tables after performing queries which change `boolean` columns to `varchar`.

Hive sample DDL query:
```
ALTER TABLE mytable CHANGE COLUMN mycolumn mycolumn varchar(5);
```

## Release notes

( ) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Hive
* Add boolean to varchar coercion for hive. ({issue}`issuenumber`)
```

### CODE REVIEW

**Review by homar — APPROVED** [review]

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestHiveCoercionOnUnpartitionedTable.java [hunk 1]
```diff
@@ -48,57 +48,58 @@ private static HiveTableDefinition.HiveTableDefinitionBuilder tableDefinitionBui
...
-                            char_to_bigger_char                CHAR(3),
-                            char_to_smaller_char               CHAR(3),
-                            timestamp_to_string                TIMESTAMP,
-                            timestamp_to_bounded_varchar       TIMESTAMP,
-                            timestamp_to_smaller_varchar       TIMESTAMP,
-                            smaller_varchar_to_timestamp       VARCHAR(4),
-                            varchar_to_timestamp               STRING,
-                            id                                 BIGINT)
-                       STORED AS\s""" + fileFormat);
+                         CREATE TABLE %NAME%(
```
[the remark this answers is not in the document]
Comment (findinpath) [reply testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestHiveCoercionOnUnpartitionedTable.java:51]: It is only `boolean_to_varchar BOOLEAN,` added (check via CLI / IDE). Apparently Github is doing diff not per line, but per value

[the remark this answers is not in the document]
Comment (ebyhr) [reply testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestHiveCoercionOnUnpartitionedTable.java:51]: No, an extra space was added to each line. Please remove them.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/coercions/BooleanCoercer.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        public BooleanToVarcharCoercer(VarcharType toType)
+        {
+            super(BOOLEAN, toType);
+        }
+
+        @Override
+        protected void applyCoercedValue(BlockBuilder blockBuilder, Block block, int position)
+        {
+            boolean value = BOOLEAN.getBoolean(block, position);
+            Slice converted = Slices.utf8Slice(Boolean.toString(value).toUpperCase(ENGLISH));
```
Comment (ebyhr) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/coercions/BooleanCoercer.java:46]: These casts (boolean → string → slice) look redundant to me. The value takes only "true" or "false".

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/coercions/CoercionUtils.java [hunk 3]
```diff
@@ -115,6 +117,9 @@ public static Type createTypeFromCoercer(TypeManager typeManager, HiveType fromH
         if (fromType instanceof VarcharType fromVarcharType && toType instanceof DateType toDateType) {
             return Optional.of(new VarcharToDateCoercer(fromVarcharType, toDateType));
         }
+        if (fromType instanceof BooleanType && toType instanceof VarcharType toVarcharType) {
+            return Optional.of(new BooleanToVarcharCoercer(toVarcharType));
```
Comment (ebyhr) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/coercions/CoercionUtils.java:121]: Should we return Optional.empty when the varchar length < 4? Sorry, I'm not familiar with this class. Asking this question as other conditions look returning the empty object instead of throwing an exception.

Comment (findinpath) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/coercions/CoercionUtils.java:121]: In such situations, the code base fails (instead of silently returning empty) with a `TrinoException` 

```
throw new TrinoException(INVALID_ARGUMENTS, format("Varchar representation of %s exceeds %s bounds", value, toType));
```

Comment (ebyhr) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/coercions/CoercionUtils.java:121]: My question was why an exception is better than empty. Other conditions seem returning empty like:
```java
        if (fromType instanceof CharType fromCharType && toType instanceof CharType toCharType) {
            if (narrowerThan(toCharType, fromCharType)) {
                return Optional.of(new CharCoercer(fromCharType, toCharType));
            }
            return Optional.empty();
        }
```

Comment (findinpath) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/coercions/CoercionUtils.java:121]: I'm not familiar enough with the functionality, but I'm assuming we should fail early instead of doing truncation like Hive does:

```
0: jdbc:hive2://localhost:10000/default> select * from testboolean;
+----------------+
| testboolean.b  |
+----------------+
| true           |
| fals           |
| NULL           |
```


cc @Praveen2112 @homar

Comment (findinpath) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/coercions/CoercionUtils.java:121]: Actually I just stumbled on `timestamp` to `varchar` coercion where Trino does truncate as well 

https://github.com/trinodb/trino/blob/e77b84893658dd422df9677bb3b3e5b9ccad7cd9/plugin/trino-hive/src/test/java/io/trino/plugin/hive/coercions/TestTimestampCoercer.java#L74-L93

I'm a bit confused now - I know the point of discussion is a corner case, but I'd like to have a clear understanding on this topic.

cc @Praveen2112

Comment (homar) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/coercions/CoercionUtils.java:121]: @findepi

Comment (Praveen2112) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/coercions/CoercionUtils.java:121]: I think we can throw an explicit exception in case of lower bounded varchar.

**Comment by Praveen2112** [comment]
Implemented by https://github.com/trinodb/trino/pull/19571
